design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -0,0 +1,286 @@
+
+We generate `Handler` implementations for said closures in [ServerOperationHandlerGenerator.kt](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerOperationHandlerGenerator.kt), for `Operation0` these are:
+
+```rust
+impl<Fun, Fut> Handler<(), Operation0Input> for Fun
+where
+    Fun: FnOnce(crate::input::Operation0Input) -> Fut,
+    Fut: Future<Output = Operation0Output>,
+{
+    async fn call(self, request: http::Request) -> http::Response {
+        let input = /* Create `Operation0Input` from `request: http::Request` */;
+
+        // Use closure on the input
+        let output = self(input).await;
+
+        let response = /* Create `http::Response` from `output: Operation0Output` */
+        response
+    }
+}
+
+impl<Fun, Fut> Handler<Extension<S>, Operation0Input> for Fun
+where
+    Fun: FnOnce(Operation0Input, Extension<S>) -> Fut,
+    Fut: Future<Output = Operation0Output>,
+{
+    async fn call(self, request: http::Request) -> http::Response {
+        let input = /* Create `Operation0Input` from `request: http::Request` */;
+
+        // Use closure on the input and fetched extension data
+        let extension = Extension(request.extensions().get::<T>().clone());
+        let output = self(input, extension).await;
+
+        let response = /* Create `http::Response` from `output: Operation0Output` */
+        response
+    }
+}
+```
+
+Creating `Operation0Input` from a `http::Request` and `http::Response` from a `Operation0Output` involves the HTTP binding traits and protocol aware deserialization. The structure [RuntimeError](https://github.com/awslabs/smithy-rs/blob/458eeb63b95e6e1e26de0858457adbc0b39cbe4e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L53-L5) enumerates internal error cases such as serialization/deserialization failures, `extensions().get::<T>()` failures, etc. We omit error handling in the snippets above, but, in full, they also involve protocol aware conversions from the `RuntimeError` and `http::Response`. The reader should make note of the influence of the model/protocol on the different sections of this procedure.
+
+The `request.extensions().get::<T>()` present in the `Fun: FnOnce(Operation0Input, Extension<S>) -> Fut` implementation is the current approach to injecting state into handlers. The customer is required to apply a [AddExtensionLayer](https://docs.rs/tower-http/latest/tower_http/add_extension/struct.AddExtensionLayer.html) to the output of the service builder so that, when the request reaches the handler, the `extensions().get::<T>()` will succeed.
```

Comment:
I'd call out in this paragraph that otherwise, the application panics, to better support your proposal.

Reply:
My comment only applies to the second paragraph; I must have mistakenly selected both in the UI when writing it.

Reply:
Actually, is this RFC going to address this extensions "type unsafety" issue? I thought you wanted to fix it but I don't see a relevant heading under the Proposal section.

Reply:
Will add a section about :+1: